### DISCUSSION TITLE
Is feature parity between MapLibre GL JS and MapLibre Native important for you?

### DISCUSSION THREAD

**Discussion by wipfli** [maplibre/maplibre#346] [General]
MapLibre has two core rendering engines: MapLibre GL JS for websites and MapLibre Native for mobile apps and more. These two toolkits are based on different programming languages and have separate implementation but both of them create near-identical maps. This means that style designers can create maps for websites and use the exact same style sheet for their mobile apps.

So far, it was a goal to keep the two toolkits at feature parity. But one may ask if this is really still an important goal given that the feature set is mature and covers a wide range of cartographic needs.

I would like to therefore ask the general question: Is feature parity between MapLibre GL JS and MapLibre Native important for you? And if yes, why?

**Comment by tyrauber** [comment]
Great question, thanks for asking!

Feature parity **and API parity** between JS and Native is extremely important.   Or rather, between `maplibre-gl-js` and `maplibre-react-native`.

Because: the line between web and native applications in react-native is quickly being erased.

Expo is making a huge push right now towards full web support for react-native, with expo-router, html-entities, native css and tailwind support (nativewind), etc.  Expo just added a server api, and will soon have some form of server side rendering.

Currently, the API between MapLibre GL JS and MapLibre Native are similar enough, that you can effectively have one JS code base for mobile and web, and merely switch the library import with a Platform conditional.  The only difficult part now is not all features supported on maplibre-gl-js exist on maplibre-native and maplibre-react-native.  This can be a real issue when you want to use a plugin across platforms, but it is only supported on GL. For example, the maplibre-contour plugin.

If the features and api continue to diverge between platforms, it means a translation layer will have to be written client-side to interface between the different APIs and features across the different maplibre libraries.

This is primarily an issue for `maplibre-gl-js` and `maplibre-react-native` where the library can and should be used for both web and mobile.

One solution, that would allow js and native to diverge further, would be to build the translation layer in `maplibre-react-native,` but that would mean that `maplibre-react-native` would only be able to support features that are available across all libraries.

This is actually a critical issue for myself, my company and my clients, who chose react-native specifically for the purpose of having one code base for all platforms.  We chose MapLibre, not only for open source, self-hosted maps, but because the api was close enough across platforms, that we wouldn't need custom map implementations for web and mobile.

This is something that I deal with everyday, so I'd be happy to have a further conversation about what the issues are here and what are the potential solutions.

**Reply by HarelM** [reply]
While I think API parity is a nice concept, I don't think it should be baked into the core repositories when I think that architecture-wise `maplibre-react-native` should be the entity that translate between the APIs as there are things that can be done easily in one platform and is hard/almost impossible to do in the other.
If you and your and your company are relaying on `maplibre-react-native` and you want to use as many features as possible in all the platforms I suggest taking the time and helping implementing those :-)
If you want to push forward maplibre-countour plugin, for example, you have two options: 
1. Open an issue in the spec repo and explain why this should be a part of the core library (I also think this might be a great addition to the spec) 
2. Open a PR in malibre-native to allow using a similar plugin that is written especially for native.

You can also invest time in `maplibre-rs` which is something that might allow what you need more easily in the far future.
Another option is to review PR and propose PR which allows for API parity that will help `maplibre-react-native`.

Bottom line, I don't oppose to the concept of API parity, but I wouldn't want to couple the API between the libraries. 
And if this is something that is relevant for you, your company and your clients, the best way to make sure it's happening is to invest your time in it :-)

**Reply by wipfli** [reply]
I wonder if people make a distinction between parity of APIs and parity of style spec features? An example for an API could be `setStyle()` and an example of a style spec feature could be the Terrain3D layer. 

@HarelM and @tyrauber do you talk about API or style spec parity?

**Reply by HarelM** [reply]
This sub-thread is about API parity, your initial thread is about spec parity.

The personas that are relevant for each type of parity are different I believe, so I think they make the distinction.
spec parity is for people designing the maps and how they would look.
API parity is for developers that are using maplibre in multiple platforms.
While there are people that are doing both, I believe it to be rare.

**Reply by tyrauber** [reply]
Yeah, I absolutely agree that feature / api parity shouldn't hold back development on any given platform. Some platforms, especially native, provide native functionality that just can't be replicated on others.  Maintainers should be free to make the most of the functionality available on their platform.  That's how the the libraries stay competitive.

And, I 100% agree with @HarelM, that the majority of the platform disparities between features and api, can and should be resolved in 'translation' libraries like maplibre-react-native.

I just think coordination between maintainers to *try* and maintain the same features and api across libraries, whenever possible,  should be a goal, because doing so greatly increases developer satisfaction and maplibre adoption.

The style-spec is a great example of this coordination.

The biggest disparity I see currently is the lack of a plugin architecture on maplibre-native.

It would be amazing it we could figure out a Plugin API that was cross-platform and worked in maplibre-native, maplibre-react-native and maplibre-gl-gs, so developers could build one plugin that would work on all platforms.  I understand this is extremely difficult, especially when taking account the different methods of handling threading across platforms. But if those differences could be abstracted in an api, it would enable developers to create cross-platform plugins which would be really, really cool.

Just my two cents. Hope this clarifies my opinion. Appreciate you all opening this question up for discussion.

**Reply by HarelM** [reply]
I haven't seen any "good" example of cross platform plugin system that you write in one language. RN, Cordova, Capacitor, Flutter , in all of them need you to write the specific code in the relevant platform (iOS, Android, Web) in order to have a single API that wraps your code in the relevant framework.
So I don't see us going in that direction, much like the fact that you can't write a plugin that will work for RN, Flutter and Capacitor - they are different. Of course you might share code, but you still need to write for each one.
But, it might be my shortsighted vision talking here, I'm I might be completely off :-)

**Reply by sjg-wdw** [reply]
There are a couple of levels here that make it confusing, but both of them are good ideas.

Down at the very lowest platform level it would be helpful to add breakaway modules.  I don't want to say plugins, because there really aren't plugins in the sense of dynamic libraries optionally loaded on mobile (so many caveats to that).  

I'd propose "Breakaway Modules" because it works like this.  Need the expanded language support?  Well, you link in this optional library, which hooks itself into MapLibre and there you go.  Does MapLibre maintain that other library... probably have to, but it's optional to link in.

That'll require a fair bit of work to expose the internal interfaces and everything you want to make breakaway has to be structured in a certain way.  On Native anyway, it's well worth doing.  JS has its own approach, also worthwhile.

But then there's the very high level.  Perhaps you want the same support for manipulating style sheets via a UI on all platforms.  You could imagine writing that in React and having it work everywhere and it makes sense since it's UI.  To do that you need to expose the same stuff on each platform in much the same way.

The first problem is core.  Some form of breakaway modules needs to happen to avoid bloating the toolkit.  The second is something you want to make easier so you get some interesting contributions from developers.

It is true that higher-ups will confuse all of this and think they can get Vulkan support or whatever using React.... but that's not really React's fault.  You're going to have a soul draining discussion about something else with that person anyway.

**Reply by tyrauber** [reply]
@sjg-wdw, @HarelM, thank you very much for your detailed responses on this thread. It's been very helpful. I think we are getting to the root of the issue.

Maintaining feature parity between maplibre-gl-js and maplibre-native is crazy difficult, and expensive.

While developing feature parity between maplibre-gl-js and maplibre-react-native is simple in comparison.  Both are Javascript. Developing a single JS API for plugins that take the platform specific details into account (caching, threads, etc) will be far less of a challenge than developing a single shared API for both native and js.

Maybe native feature parity isn't required?  Maybe people building mobile apps in Swift, Objective-C or Kotlin, don't need Plugins? Maybe they just need maplibre-native as a low-level rendering engine?  

Is style-spec parity and rendering consistency all that matters?

Hopefully someone that uses maplibre-native in production can weigh in here, but the answer seems pretty clear to me now.

**Reply by sjg-wdw** [reply]
> Developing a single JS API for plugins that take the platform specific details into account (caching, threads, etc) will be far less of a challenge than developing a single shared API for both native and js.

Oh my!  No.  This would be incredibly difficult to get working right and perform terribly on the mobile platforms.

I understand where the confusion comes from.  I've had versions of this conversation many times over the years with mobile projects.

Think of React, Flutter and all the toolkits like them in the past (and yet to come) as friendly wrappers on top of core functionality.  They're not meant to do difficult, performant things like a full-on map toolkit.  They're for knitting your UI together.
Once you start thinking of replacing fairly low level services, like tile caching or rendering, with a new implementation using their constructs you're off their main path.

**Reply by tyrauber** [reply]
🤣 Ha, ok then.  Is feature parity between MapLibre GL JS and MapLibre Native important for you? Yes.

Or more accurately, between maplibre-react-native and maplibre-gl-js. And because maplibre-react-native depends so heavily on maplibre-native, between maplibre-native and maplibre-gl-js.

maplibre-react-native **needs** an api from maplibre-native to leverage the low level rendering engine in such a way that it can provide feature parity with maplibre-gl-js. Otherwise the feature set used for cross-platform apps will be the lowest common denominator.

**Comment by sjg-wdw** [comment]
There are several kinds of feature parity to consider:
- Style Sheet Feature Parity.
- Android/iOS/Linux/Windows Feature Parity
- Native & Web Feature Parity
- React/Flutter Parity

Parity is a good ideal, but it's easy to get too focused and spend money on things no one cares about.  Let me throw out these categories of how much I would recommend caring:
- Yes, definitely
- Sure, someday
- Eh, nah.

## Style Sheet Feature Parity

This has the best argument for parity across all platforms.  If MapLibre is the keeper of the style sheet spec, then the various versions should probably support everything, where possible.

Also cartographers have this habit of developing their styles on the web and then expecting it to work for mobile.

Category: Yes, definitely.

## Intra-Native Feature Parity

Native consists of two mobile platforms (iOS & Android), NodeJS & Linux, Qt, and then the long tail of Windows and anything else someone has gotten running.

Style sheet parity is the last section and that's a "Yes, definitely".  You want your offline rendering to look like your web & mobile rendering.  That's kind of the point.

As for API parity, that's a bit more complicated.  Parity between iOS & Android kind of makes sense.... although those platforms work differently.  You might want to implement things that only work on iOS (e.g. SceneKit interfacing) without having to worry about Android and vice versa.

Category: Sure, someday.

Parity between the mobile platforms and the other Native platforms is even odder.  Do you need an annotation API for NodeJS to use?

Category: Eh, nah.

## Native & Web Feature Parity

There's something to be said for parity between web and mobile, at least for the standard map display.  You want the style sheets there on both and you want it to work more or less the same way, at least within the context of the platform itself.  Making things work exactly the same across all platforms is the next section.

But some things are always going to be very different.  Offline support?  Would be good if both mobile platforms worked similarly, but do you even want to try this for the web?  Probably not.

Other things, like terrain, you may implement for one and then decide you'd like for the other.  A good goal, but don't hold development back.  Let the developers run and then play catchup later.

Category: Sure, someday

## React/Flutter Parity

These are the current "all the same, everywhere" platforms and they're worth interfacing to, but not letting them hold you back.  In my experience the developers of these sorts of toolkits are great to work with, but their users don't really see you.  

The point of these is to smooth out the rough terrain of mobile & web development, so help them hook up functionality.  Maybe figure out what level that makes sense at and define a core set of features you want to be available.  But don't worry about keeping them on the bleeding edge of new development.

Category: Sure, someday

**Comment by danielgard** [comment]
At [komoot](https://www.komoot.com) we are using MapLibre on Web and Mobile with native platform implementations. 

We’ve chosen MapLibre because of its fundamental concept to have different SDKs that can be used with all the native advantages on each major platform to be able to display the same map. With that I won’t catch up on API or other parities here. 

In general I agree a lot with what @sjg-wdw said in his earlier post. This is our point of few about Style Sheet Feature Parity:

Having one flexible source for changing the look and feel of the map on all platforms - the style - is for us, and I believe for many others, the key aspect of MapLibre. If that wouldn’t be there we probably wouldn’t choose MapLibre as our map SDKs.

It enables us to have a unified user experience and to iterate quickly on map details. People who don’t have a developer background - but a design and maps background - continuously improve the look and feel of our map by changing our map style. Map based features of our app always take into account the possibilities of MapLibre styling options. It’s great to focus on whats possible and use the options provided to build a unified user experience on all platforms instead of having different user experiences on different platforms. 

--- 

**To answer the question in short:**

Is feature parity between MapLibre GL JS and MapLibre Native important for you? 
_YES_

And if yes, why?
_It’s important for us to build a unified user experience with our map and to be able to iterate quickly and often on map details._
